src/System.IO.FileSystem/src/System/IO/Win32FileStreamAsyncResult.cs

```diff
@@ -96,232 +53,147 @@ unsafe private sealed class FileStreamAsyncResult : IAsyncResult
+                            Monitor.Exit(_completionSource);
+                        }
+                    }
                 }
             }
 
-#if USE_OVERLAPPED
             internal void ReleaseNativeResource()
             {
-                // Free memory & GC handles.
-                if (this._overlapped != null)
-                    Overlapped.Free(_overlapped);
-            }
-#endif
-
-            internal void Wait()
-            {
-                if (_waitHandle != null)
+                // Lock to prevent race between registering the cancellation and cleaning the overlapped
+                lock (_completionSource)
                 {
-                    // We must block to ensure that AsyncFSCallback has completed,
-                    // and we should close the WaitHandle in here.  AsyncFSCallback
-                    // and the hand-ported imitation version in COMThreadPool.cpp 
-                    // are the only places that set this event.
-                    try
-                    {
-                        _waitHandle.WaitOne();
-                        Debug.Assert(_isComplete == true, "FileStreamAsyncResult::Wait - AsyncFSCallback  didn't set _isComplete to true!");
-                    }
-                    finally
+                    // Ensure that cancellation has been completed and cleaned up
+                    _cancellationRegistration.Dispose();
+
+                    // Free the overlapped
+                    // NOTE: The cancellation must *NOT* be running at this point, or it may observe freed memory
+                    // (this is why we disposed the registration above)
+                    if (_overlapped != null)
                     {
-                        _waitHandle.Dispose();
+                        Threading.Overlapped.Free(_overlapped);
+                        _overlapped = null;
                     }
                 }
             }
-#if USE_OVERLAPPED
+
             // When doing IO asynchronously (ie, _isAsync==true), this callback is 
             // called by a free thread in the threadpool when the IO operation 
             // completes.  
             unsafe private static void AsyncFSCallback(uint errorCode, uint numBytes, NativeOverlapped* pOverlapped)
             {
                 // Unpack overlapped
-                Overlapped overlapped = Overlapped.Unpack(pOverlapped);
-                // Free the overlapped struct in EndRead/EndWrite.
+                Overlapped overlapped = Threading.Overlapped.Unpack(pOverlapped);
 
-                // Extract async result from overlapped 
-                FileStreamAsyncResult asyncResult =
-                    (FileStreamAsyncResult)overlapped.AsyncResult;
-                asyncResult._numBytes = (int)numBytes;
+                // Extract async result from overlapped, and then free the native overlapped
+                FileStreamAsyncResult overlappedToTask = (FileStreamAsyncResult)overlapped.AsyncResult;
```

Comment:
Ah, is this why you kept the type implementing IAsyncResult?  Too bad the overlapped type has its state strongly typed like that.